code/controllers/subsystem/autofire.dm

```diff
@@ -0,0 +1,217 @@
+		reset_buckets()
+		bucket_list = src.bucket_list
+		resumed = FALSE
+
+	// Store a reference to the 'working' shooter so that we can resume if the MC
+	// has us stop mid-way through processing
+	var/static/datum/component/automatedfire/shooter
+	if (!resumed)
+		shooter = null
+
+	// Iterate through each bucket starting from the practical offset
+	while (practical_offset <= BUCKET_LEN && head_offset + ((practical_offset - 1) * world.tick_lag) <= world.time)	
+		if(!shooter)
+			shooter =  bucket_list[practical_offset]
+			bucket_list[practical_offset] = null
+
+		while (shooter)
+			next_shooter = shooter.next
+			INVOKE_ASYNC(shooter, /datum/component/automatedfire/proc/process_shot)
+
+			SSautomatedfire.bucket_count--
+			shooter = next_shooter
+			if (MC_TICK_CHECK)
+				return
+
+		// Empty the bucket
+		practical_offset++
+
+/datum/controller/subsystem/automatedfire/Recover()
+	bucket_list |= SSautomatedfire.bucket_list
+
+///In the event of a change of world.tick_lag, we refresh the size of the bucket and the bucket resolution
+/datum/controller/subsystem/automatedfire/proc/reset_buckets()
+	bucket_list.len = BUCKET_LEN
+	head_offset = world.time
+	bucket_resolution = world.tick_lag
+
+/datum/component/automatedfire
+	///The owner of this component
+	var/atom/shooter
+	/// Contains the scheduled fire time, used for scheduling EOL
+	var/next_fire
+	/// Contains the reference to the next component in the bucket, used by autofire subsystem
+	var/datum/component/automatedfire/next
+	/// Contains the reference to the previous component in the bucket, used by autofire subsystem
+	var/datum/component/automatedfire/prev
+
+
+/// chedule the shooter into the system, inserting it into the next fire queue
+/datum/component/automatedfire/proc/schedule_shot()
+	//We move to another bucket, so we clean the reference from the former linked list
+	next = null
+	prev = null	
+	var/list/bucket_list = SSautomatedfire.bucket_list
+
+	// Ensure the next_fire time is properly bound to avoid missing a scheduled event
+	next_fire = max(CEILING(next_fire, world.tick_lag), world.time + world.tick_lag)
+
+	// Get bucket position and a local reference to the datum var, it's faster to access this way
+	var/bucket_pos = BUCKET_POS(next_fire)
+
+	// Get the bucket head for that bucket, increment the bucket count
+	var/datum/component/automatedfire/bucket_head = bucket_list[bucket_pos]
+	SSautomatedfire.bucket_count++
+
+	// If there is no existing head of this bucket, we can set this shooter to be that head
+	if (!bucket_head)
+		bucket_list[bucket_pos] = src
+		return
+
+	// Otherwise it's a simple insertion into the double-linked list
+	if (bucket_head.next)
+		next = bucket_head.next
+		next.prev = src
+
+	bucket_head.next = src
+	prev = bucket_head
+
+	//Something went wrong, probably a lag spike or something. To prevent infinite loops, we reschedule it to a another next fire
+	if(prev == src)
+		next_fire = next_fire += 1
```

Comment:
Like we talked on discord, this can happen due to byondism, or bad code (like if the component is schedule twice for the same tick)

> but because you insert elements in the linked list after bucket_head, which means that they won't be processed in > that tick, because shooter is already after this element most of the time. Runechat adds new elements before head

This will never happen unless you schedule a component for the immediate tick, which is not something you should ever do. Replacing the bucket head each time like runechat does is useless for this system